File: Samples/RadioManager/cs/RadioModel.cs
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using Windows.Devices.Radios;
using Windows.UI.Core;
using Windows.UI.Xaml;

namespace SDKTemplate
{
    public class RadioModel : INotifyPropertyChanged
    {
        private Radio radio;
        private bool isEnabled;
        private CoreDispatcher dispatcher;

        public RadioModel(Radio radio, CoreDispatcher dispatcher)
        {
            this.radio = radio;
            // Controlling the mobile broadband radio requires the cellularDeviceControl restricted capability, which we do not have.
            this.isEnabled = (radio.Kind != RadioKind.MobileBroadband);
            this.dispatcher = dispatcher;
            this.radio.StateChanged += Radio_StateChanged;
        }

        private async void Radio_StateChanged(Radio sender, object args)
        {
            // The Radio StateChanged event doesn't run from the UI thread, so we must use the dispatcher
            // to run NotifyPropertyChanged
            await this.dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
            {
                NotifyPropertyChanged("IsRadioOn");
            });
        }

        public string Name
        {
            get
            {
                return this.radio.Name;
            }
        }

        public bool IsRadioOn
        {
            get
            {
                return this.radio.State == RadioState.On;
            }
            set
            {
                SetRadioState(value);
            }
        }

        public bool IsEnabled
        {
            get
            {
                return this.isEnabled;
            }
            set
            {
                this.isEnabled = value;
                NotifyPropertyChanged();
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;

        private void NotifyPropertyChanged([CallerMemberName] string propertyName = "")
        {
            if (PropertyChanged != null)
            {
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
            }
        }

        private async void SetRadioState(bool isRadioOn)
        {
            if (isRadioOn != IsRadioOn)
            {
                var radioState = isRadioOn ? RadioState.On : RadioState.Off;
                IsEnabled = false;
                await this.radio.SetStateAsync(radioState);
                NotifyPropertyChanged("IsRadioOn");
                IsEnabled = true;
            }
        }
    }
}

File: Samples/RadioManager/cppwinrt/RadioModel.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include <winrt/SDKTemplate.h>
#include "RadioModel.h"
#include "RadioModel.g.cpp"

using namespace winrt;
using namespace winrt::Windows::Devices::Radios;
using namespace winrt::Windows::Foundation::Collections;
using namespace winrt::Windows::UI::Core;
using namespace winrt::Windows::UI::Xaml::Data;
using namespace winrt::SDKTemplate;

namespace winrt::SDKTemplate::implementation
{
    RadioModel::RadioModel(Radio const& radio, CoreDispatcher const& dispatcher) :
        radio(radio), dispatcher(dispatcher),
        // Controlling the mobile broadband radio requires the cellularDeviceControl restricted capability, which we do not have.
        isEnabled(radio.Kind() != RadioKind::MobileBroadband)
    {
        radio.StateChanged({ get_weak(), &RadioModel::Radio_StateChanged });
    }

    fire_and_forget RadioModel::Radio_StateChanged(Radio const&, IInspectable const&)
    {
        // The Radio StateChanged event doesn't run from the UI thread, so we must use the dispatcher
        // to run NotifyPropertyChanged
        auto lifetime = get_strong();
        co_await resume_foreground(dispatcher);
        NotifyPropertyChanged(L"IsRadioOn");
    }

    void RadioModel::NotifyPropertyChanged(hstring const& propertyName)
    {
        propertyChanged(*this, PropertyChangedEventArgs(propertyName));
    }

    void RadioModel::SetEnabled(bool value)
    {
        isEnabled = value;
        NotifyPropertyChanged(L"IsEnabled");
    }

    fire_and_forget RadioModel::SetRadioState(bool value)
    {
        if (value != IsRadioOn())
        {
            auto lifetime = get_strong();
            RadioState newState = value ? RadioState::On : RadioState::Off;
            SetEnabled(false);
            co_await radio.SetStateAsync(newState);
            SetEnabled(true);
            NotifyPropertyChanged(L"IsRadioOn");
        }
    }

}
